File: Problem1-queueUsingStack.py
# -*- coding: utf-8 -*-
class MyQueue:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.list1 = []
        self.list2 = []
        

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self.list1.append(x)
        #print("push ",x)
        #print(self.list1,self.list2)
        

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        if not self.list2:
            self.list2 = self.list1[::-1]
            self.list1 = []
        #print("pop")
        #print(self.list1,self.list2)
        #print("pooped ",self.list2[-1]) 
        return self.list2.pop()
            
        

    def peek(self) -> int:
        """
        Get the front element.
        """
        if not self.list2:
          #  print("peek1")
          #  print(self.list1,self.list2,self.list1[0])
            return self.list1[0]
        else:
          #  print("peek2")
         #   print(self.list1,self.list2,self.list2.peek())
            return self.list2[-1]
        

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
       # print(not self.list1 and not self.list2)
        if (not self.list1) and not self.list2:
            return True
        else:
            return False


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: Problem1_QueueUsingStack.cs
// Approach : Use two stacks
// push to first one
// when popping or peeking put the elements to second stack to reverse the order and make them FIFO

public class MyQueue {
    
    Stack<int> inStack, outStack;

    public MyQueue() {
        inStack = new Stack<int>();
        outStack = new Stack<int>();
    }
    
    private void transferIfNeeded(){
        if(outStack.Count == 0){
            while(inStack.Count !=0)
                outStack.Push(inStack.Pop());
        }
    }
    
    // Time Complexity: O(1)
    // Space Complexity: O(1)
    public void Push(int x) {
        inStack.Push(x);
    }
    
    // Amortized Time Complexity: O(1)
    // Space Complexity: O(1)
    public int Pop() {
        transferIfNeeded();
        return outStack.Pop();
    }
    
    // Amortized Time Complexity: O(1)
    // Space Complexity: O(1)
    public int Peek() {
        transferIfNeeded();
        return outStack.Peek();
    }
    
    // Time Complexity: O(1)
    // Space Complexity: O(1)
    public bool Empty() {
        return inStack.Count==0 && outStack.Count==0;
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.Push(x);
 * int param_2 = obj.Pop();
 * int param_3 = obj.Peek();
 * bool param_4 = obj.Empty();
 */